lib/jxl/dec_frame.cc

```diff
@@ -827,6 +827,63 @@ Status FrameDecoder::Flush() {
   return true;
 }
 
+int FrameDecoder::ReferencedBy() const {
+  if (is_finalized_) {
+    // header not parsed
+    return 0;
+  }
+  if (frame_header_.frame_type == FrameType::kDCFrame) {
+    // bits 16, 32, 64, 128 for DC level
+    return 16 << (frame_header_.dc_level - 1);
+  } else if (frame_header_.CanBeReferenced()) {
+    // bits 1, 2, 4 and 8 for the references
+    return 1 << frame_header_.save_as_reference;
+  }
+
+  return 0;
+}
+
+int FrameDecoder::References() const {
+  if (is_finalized_) {
+    return 0;
+  }
+  if ((!decoded_dc_global_ || !decoded_ac_global_ ||
+       *std::min_element(decoded_dc_groups_.begin(),
+                         decoded_dc_groups_.end()) != 1 ||
+       *std::min_element(decoded_passes_per_ac_group_.begin(),
+                         decoded_passes_per_ac_group_.end()) < max_passes_)) {
+    return 0;
+  }
+
+  int result = 0;
+
+  // Blending
+  if (frame_header_.blending_info.mode != BlendMode::kReplace) {
```

Comment:
Also if it is kReplace, there can be a dependency if the frame is not a full frame but a cropped one.

Reply:
Is this the case as soon as `custom_size_or_origin` is true, or only if `custom_size_or_origin` is true and the frame_size and frame_origin are an actual crop?

Reply:
Second one - but first one likely works too...

There's also a NeedsBlending() method somewhere.

Reply:
Thanks for the pointer! Used NeedsBlending's implementation to know what to write here, did not actually use NeedsBlending directly from here though because it requires implementing the same loop anyway to know the sources of the extra channels dependencies